File: yastd/lib/yast2/software.rb
# frozen_string_literal: true

# Copyright (c) [2021] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "yast"
require "yast2/package_callbacks"
require "y2packager/product"

Yast.import "Pkg"
Yast.import "PackageInstallation"
Yast.import "Stage"

# YaST specific code lives under this namespace
module Yast2
  # This class is responsible for software handling
  class Software
    attr_reader :product, :products

    def initialize(logger)
      @logger = logger
      @products = []
      @product = nil
    end

    def select_product(name)
      raise ArgumentError unless @products.any? { |p| p.name == name }

      @product = name
    end

    def probe
      logger.info "Probing software"
      # as we use liveDVD with normal like ENV, lets temporary switch to normal to use its repos
      Yast::Stage.Set("normal")
      Yast::Pkg.TargetInitialize("/")
      Yast::Pkg.TargetLoad
      Yast::Pkg.SourceRestore
      Yast::Pkg.SourceLoad
      @products = Y2Packager::Product.all
      @product = @products.first&.name
      proposal = Yast::Packages.Proposal(force_reset = true, reinit = false, _simple = true)
      @logger.info "proposal #{proposal["raw_proposal"]}"
      Yast::Stage.Set("initial")

      raise "No Product Available" unless @product
    end

    def propose
      selected_product = @products.find { |p| p.name == @product }
      selected_product.select
      @logger.info "selected product #{selected_product.inspect}"

      # as we use liveDVD with normal like ENV, lets temporary switch to normal to use its repos
      Yast::Stage.Set("normal")
      proposal = Yast::Packages.Proposal(force_reset = false, reinit = false, _simple = true)
      @logger.info "proposal #{proposal["raw_proposal"]}"
      res = Yast::Pkg.PkgSolve(unused = true)
      @logger.info "solver run #{res.inspect}"

      Yast::Stage.Set("initial")
      # do not return proposal hash, so intentional nil here
      nil
    end

    def install(progress)
      count_packages(progress)

      PackageCallbacks.setup(progress)

      # TODO: error handling
      Yast::Pkg.TargetInitialize(Yast::Installation.destdir)
      commit_result = Yast::PackageInstallation.Commit({})

      if commit_result.nil? || commit_result.empty?
        log.error("Commit failed")
        raise Yast::Pkg.LastError
      end
    end

  private

    attr_reader :logger

    def count_packages(progress)
      count = Yast::Pkg.PkgMediaCount.reduce(0) { |sum, res| sum + res.reduce(0, :+) }

      progress.packages_to_install = count
    end
  end
end

File: web/src/model/config/software.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

/**
 * List of user-selected patterns to install
 */
export type PatternsArray = string[];

/**
 * Software configuration.
 */
export interface Config {
  patterns?: PatternsArray | PatternsObject;
  /**
   * List of packages to install
   */
  packages?: string[];
  /**
   * Flag if only minimal hard dependencies should be used in solver
   */
  onlyRequired?: boolean;
  /**
   * List of user specified repositories that will be used on top of default ones
   */
  extraRepositories?: Repository[];
  [k: string]: unknown;
}
/**
 * Modifications for the list of user-selected patterns to install
 */
export interface PatternsObject {
  /**
   * List of user-selected patterns to add to the list
   */
  add?: string[];
  /**
   * List of user-selected patterns to remove from the list
   */
  remove?: string[];
}
/**
 * Packages repository
 */
export interface Repository {
  /**
   * alias used for repository. Acting as identifier
   */
  alias?: string;
  /**
   * URL pointing to repository
   */
  url?: string;
  /**
   * Repository priority
   */
  priority?: number;
  /**
   * User visible name. Defaults to alias
   */
  name?: string;
  /**
   * product directory on multi repo DVD. Usually not needed
   */
  productDir?: string;
  /**
   * If repository should be enabled. Defaults to true. Useful when adding additional repo that should not be immediately use.
   */
  enabled?: boolean;
  /**
   * If unsigned repositories are allowed. Mainly useful for repositories that is hand crafted without GPG signature.
   */
  allowUnsigned?: boolean;
  /**
   * List of GPG fingerprints that is accepted for this repository. Useful for own repositories with proper GPG signature.
   */
  gpgFingerprints?: string[];
}
